### PR TITLE
Add lease APIs

### PR DESCRIPTION
**Pull request by mmihic**
Along with fakes for cluster and lease management

### CODE REVIEW

File: misk/src/main/kotlin/misk/clustering/lease/Lease.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * flapping, service lease owners may want to delay releasing leases until they've been not ready
+ * for a particular amount of time.
+ */
+interface Lease {
+  /** @property String the name of the resource being leased */
+  val name: String;
+
+  /**
+   * @return true if the lease is owned by this process instance. This may involve remote calls,
+   * so it is marked as a function rather than a property to make the potential expense clearer
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/clustering/lease/Lease.kt]: 👍🏻

File: misk/src/main/kotlin/misk/clustering/lease/LeaseManager.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+interface LeaseManager {
+  /**
+   * Registers interest in the lease with the given name. Service instances should register their
+   * interest in leases as soon as they know about them; this gives the process an opportunity
+   * to notify the underlying cluster system that it is interested in the lease, or to setup
+   * other background tasks to acquire the lease if necessary. Note that registering interest
+   * in a lease doesn't necessarily mean that this service instance will attempt to acquire
+   * the lease; many clustering systems will only try to acquire leases for resources that
+   * they think they should own based on some consistent hashing system
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/clustering/lease/LeaseManager.kt]: nit: missed a `.`

also https://nedbatchelder.com/blog/201401/comments_should_be_sentences.html

File: misk/src/main/kotlin/misk/clustering/lease/LeaseManager.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+   */
+  fun requestLease(name: String): Lease
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/clustering/lease/LeaseManager.kt]: I’m not sure on this API. As a lease user I’m expecting methods like `acquire()` and `release()`, this API has a weaker `acquire()` and no `release()`.

Should we call this something else?

Comment (mmihic) [reply misk/src/main/kotlin/misk/clustering/lease/LeaseManager.kt]: That's intentional. You don't explicitly `acquire` and `release` a lease - you register interest in it, and if the clustering system decides you are the instance that it should be assigned to, then in the background it is acquired. We're modeling load balanced leases here, where all processes in the cluster indicate that they are potential candidates for owning a lease, and the actual distribution of named lease ownership is handled by the clustering system.

Comment (mmihic) [reply misk/src/main/kotlin/misk/clustering/lease/LeaseManager.kt]: That said, I'm open to other names.

Comment (mmihic) [reply misk/src/main/kotlin/misk/clustering/lease/LeaseManager.kt]: I merged with the current name, but I'm also not happy with it so if you have other suggestions, I'm down to change it.

Comment (swankjesse) [reply misk/src/main/kotlin/misk/clustering/lease/LeaseManager.kt]: Thumbs up. I don’t have better names either.

**Pull request merged**